### ISSUE TITLE
lsp: incremental lint diagnostic generation

### ISSUE THREAD

**Issue by dsherret** [denoland/deno#13578]
Right now, we create lint diagnostics for every open file each time:

https://github.com/denoland/deno/blob/ed3086e4b129843c13a009112cf21dfd05745905/cli/lsp/diagnostics.rs#L451

This means that after you open enough files (say 50), you will start to get a noticable delay when changing only one file.

To improve this, we could store lint diagnostics and only recreate them when the file version changes. This should be ok because lint diagnostics are not dependent on other files. I don't think this would work for things like TypeScript or dependency diagnostics unfortunately.

**Comment by kitsonk** [comment]
Yes, it wouldn't work for TypeScript diagnostics, but I do think think this would work for the the dependency diagnostics. We have the concept of dependants in the language server, and the only events that we would really care about is that a `cache` has changed, or a file has been created (as it may be a missing dependency of an existing file), or a dependent file has been deleted.
